qa/tasks/keystone.py

```diff
@@ -0,0 +1,405 @@
+            dict_to_args(special, auth_section + section_item.items()))
+
+@contextlib.contextmanager
+def create_users(ctx, config):
+    """
+    Create a main and an alternate s3 user.
+    """
+    assert isinstance(config, dict)
+
+    (cclient, cconfig) = config.items()[0]
+
+    # configure tenants/projects
+    run_section_cmds(ctx, cclient, 'project create', 'name',
+                     cconfig['tenants'])
+    run_section_cmds(ctx, cclient, 'user create', 'name',
+                     cconfig['users'])
+    run_section_cmds(ctx, cclient, 'role create', 'name',
+                     cconfig['roles'])
+    run_section_cmds(ctx, cclient, 'role add', 'name',
+                     cconfig['role-mappings'])
+    run_section_cmds(ctx, cclient, 'service create', 'name',
+                     cconfig['services'])
+    run_section_cmds(ctx, cclient, 'endpoint create', 'service',
+                     cconfig['endpoints'])
+
+    # sleep driven synchronization -- just in case
+    run_in_keystone_venv(ctx, cclient, [ 'sleep', '3' ])
+    try:
+        yield
+    finally:
+        pass
+
+
+@contextlib.contextmanager
+def task(ctx, config):
+    """
+    Deploy and configure Keystone
+
+    Example of configuration:
+
+      tasks:
+      - local_cluster:
+          cluster_path: /home/rzarzynski/ceph-1/build
+      - local_rgw:
+      - keystone:
+          client.0:
+            force-branch: master
+            tenants:
+              - name: admin
+                description:  Admin Tenant
+              - name: tempest
+                description: Tempest Tenant
+            users:
+              - name: admin
+                password: ADMIN
+                project: admin
+              - name: tempest_u1
+                password: tempest
+                project: tempest
+              - name: tempest_u2
+                password: tempest
+                project: tempest
+            roles: [ name: admin, name: Member ]
+            role-mappings:
+              - name: admin
+                user: admin
+                project: admin
+              - name: Member
+                user: tempest_u1
+                project: tempest
+              - name: Member
+                user: tempest_u2
+                project: tempest
+            services:
+              - name: keystone
+                type: identity
+                description: Keystone Identity Service
+              - name: swift
+                type: object-store
+                description: Swift Service
+              endpoints:
+                - service: keystone
+                  publicurl: http://127.0.1:$(public_port)s/v2.0
+                - service: swift
+                  publicurl: http://127.0.0.1:8000/v1/KEY_$(tenant_id)s
+    """
+    assert config is None or isinstance(config, list) \
+        or isinstance(config, dict), \
+        "task keystone only supports a list or dictionary for configuration"
```

Comment:
since this task depends on `ctx.tox`, you should enforce that here:
```python
    if not ctx.tox:
        raise ConfigError('keystone must run after the tox task')
```